python/cuxfilter/charts/core/non_aggregate/utils.py

```diff
@@ -1,13 +1,177 @@
+        Output boolean array. out[i] is True if points_xy[2*i:2*i+2] is
+        inside the polygon.
+    """
+    i = cuda.grid(1)  # Global thread index
+
+    # Check if index is within bounds (number of points)
+    if i < points_xy.shape[0] // 2:
+        px = points_xy[2 * i]
+        py = points_xy[2 * i + 1]
+
+        # Call the device function for the actual check
+        out[i] = point_in_polygon_ray_cast(px, py, polygon_xy, num_vertices)
+
+
+def point_in_polygon(df, x, y, polygon_coords):
+    """
+    Checks which points in a DataFrame are inside a given polygon using a CUDA
+    kernel.
+
+    Parameters
+    ----------
+    df : cudf.DataFrame
+        DataFrame containing the points.
+    x : str
+        Column name for the x-coordinates of the points.
+    y : str
+        Column name for the y-coordinates of the points.
+    polygon_coords : list of tuples or similar array-like
+        Coordinates of the polygon vertices, e.g., [(x1, y1), (x2, y2), ...].
+        The polygon should be closed.
+
+    Returns
+    -------
+    cudf.Series
+        A boolean Series indicating whether each point is inside the polygon.
+    """
+    if not isinstance(df, cudf.DataFrame):
+        raise TypeError("Input 'df' must be a cudf.DataFrame")
+
+    # 1. Prepare point data
+    num_points = len(df)
+    if num_points == 0:
+        return cudf.Series([], dtype="bool", index=df.index)
+
+    # Interleave columns directly using cudf and get cupy array
+    # .values returns a cupy array for cudf Series/DataFrame
+    points_xy_cp = df[[x, y]].astype(cp.float64).interleave_columns().values
+
+    # 2. Prepare polygon data
+    try:
+        # Directly create cupy array from polygon_coords
+        polygon_xy_cp = cp.array(polygon_coords, dtype=cp.float64)
+
+        # Flatten if it's a list of pairs (N, 2)
+        if polygon_xy_cp.ndim == 2 and polygon_xy_cp.shape[1] == 2:
+            polygon_xy_cp = polygon_xy_cp.flatten()
+        elif polygon_xy_cp.ndim != 1:
+            raise ValueError(
+                "Polygon coordinates must be list of pairs or flat list."
+            )
+
+        # Validate size
+        if polygon_xy_cp.size % 2 != 0:
+            raise ValueError(
+                "Polygon coordinates must have an even number of elements."
+            )
+
+        num_vertices = polygon_xy_cp.size // 2
+
+        if num_vertices < 3:
+            raise ValueError("Polygon must have at least 3 vertices.")
+
+    except (ValueError, TypeError):
+        # Handle invalid polygon formats or insufficient vertices
+        # Return all False as per the check at the start
+        if isinstance(df.index, cudf.MultiIndex):
+            idx = cudf.RangeIndex(len(df))
+        else:
+            idx = df.index
+        return cudf.Series([False] * len(df), index=idx, dtype="bool")
+
+    # 3. Prepare output array
+    out_gpu = cp.zeros(num_points, dtype=cp.bool_)
+
+    # 4. Configure and launch kernel
+    threads_per_block = 128
```

Comment:
Naively, 256 can be a better default here. Depends on the kernel. Please check whether performance is affected.

Reply:
Not much, I am getting similar results eitherways:

fluctuates between multiple runs:

--- Calling point_in_polygon (1st time - includes compilation) ---
First call duration: 15-25ms

--- Calling point_in_polygon (2nd time - uses compiled kernel) ---
Second call duration: 5-6ms

Reply:
updated to 256 nonetheless